### PR TITLE
Increased efficiency and speed of subscriber status refresh

### PR DESCRIPTION
**Pull request by ALazyMeme**
Pull request checklist:

- [x] `CHANGELOG.md` was updated, if applicable
- [x] Documentation in docs/ or install-docs/ was updated, if applicable
- [ ] I have tested all changes

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,7 @@ Remember to bring your dependencies up to date with `./scripts/venvinstall.sh` w
 - Breaking: Migrated Announce from IRC to Helix. (#2141)
 - Breaking: Migrated Delete moderation action from IRC to Helix. (#2173)
 - Minor: Migrated LastFM module to the `reply` response type. (#2118, #2128)
+- Dev: Improved Helix subscriber fetching to utilize the existing data provided by the Helix response. (#2203)
```
Comment (RAnders00) [inline CHANGELOG.md:18]:
```suggestion
- Minor: Increased efficiency and speed of subscriber status refresh. (#2203)
```

**Review by RAnders00 — CHANGES_REQUESTED** [review]
Looks good otherwise as far as I can tell! Thank you!

**Review by RAnders00 — APPROVED** [review]

**Comment by pajlada** [comment]
Running into this issue where the streamer can appear twice in the endpoint, not sure why this would happen since we should do a `list(set(...))`, needs some more investigation

```
Logging an uncaught exception (ActionQueue)
Traceback (most recent call last):
  File "/home/pajlada/git/pajbot/venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1880, in _execute_context
    self.dialect.do_executemany(
  File "/home/pajlada/git/pajbot/venv/lib/python3.10/site-packages/sqlalchemy/dialects/postgresql/psycopg2.py", line 1000, in do_executemany
    cursor.executemany(statement, parameters)
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "subscribers_pkey"
DETAIL:  Key (id)=(11148817) already exists.
```

**Review by pajlada — CHANGES_REQUESTED** [review]
^

**Comment by RAnders00** [comment]
I see the issue. I *assumed* we had this already because I assumed an equality comparison on two `UserBasics` objects would have already been used somewhere else in the code base, but we don‘t have equality and hash implemented for `UserBasics`.

https://github.com/pajbot/pajbot/blob/6b7681f29ce0f4bdb82d22bf44863c091c312094/pajbot/models/user.py#L42
https://stackoverflow.com/a/2909119

As a result of the hash function not being implemented, `set()` is not able to eliminate duplicates.

**Comment by RAnders00** [comment]
Also, I‘d like to argue that the fetcher function could very well return sets already. It‘s quite wasteful to turn list into set and back into list. If we directly used sets all the way (in the `_fetch_all_pages` method and in the `_fetch_X_page` method) we wouldn‘t need to do this. A conversion back to list is probably never necessary (after all, there‘s no inherent order to the list of subs.)

File: pajbot/apiwrappers/twitch/helix.py [hunk 2]
```diff
@@ -340,19 +340,21 @@ def _fetch_subscribers_page(self, broadcaster_id: str, authorization, after_pagi
...
         """Fetch a list of all subscribers (user IDs) of a broadcaster."""
-        subscriber_ids = self._fetch_all_pages(self._fetch_subscribers_page, broadcaster_id, authorization)
+        subscribers = self._fetch_all_pages(self._fetch_subscribers_page, broadcaster_id, authorization)
 
         # Dedupe the list of subscribers since the API can return the same IDs multiple times
-        subscriber_ids = list(set(subscriber_ids))
+        subscribers = list(set(subscribers))
 
-        return subscriber_ids
+        return subscribers
```
Comment (RAnders00) [inline pajbot/apiwrappers/twitch/helix.py:357]:
```suggestion
        return set(subscribers)
```

File: pajbot/apiwrappers/twitch/helix.py [hunk 3]
```diff
@@ -340,19 +340,21 @@ def _fetch_subscribers_page(self, broadcaster_id: str, authorization, after_pagi
...
         return subscribers, pagination_cursor
 
     def fetch_all_subscribers(self, broadcaster_id, authorization):
         """Fetch a list of all subscribers (user IDs) of a broadcaster."""
-        subscriber_ids = self._fetch_all_pages(self._fetch_subscribers_page, broadcaster_id, authorization)
+        subscribers = self._fetch_all_pages(self._fetch_subscribers_page, broadcaster_id, authorization)
 
         # Dedupe the list of subscribers since the API can return the same IDs multiple times
-        subscriber_ids = list(set(subscriber_ids))
+        subscribers = list(set(subscribers))
```
Comment (RAnders00) [inline pajbot/apiwrappers/twitch/helix.py:355]:
```suggestion
```

**Review by RAnders00 — APPROVED** [review]

**Comment by RAnders00** [comment]
Oops, I didn‘t mean to select „Approve“, consider it a „Changes requested“, please.

**Review by pajlada — APPROVED** [review]
Works as expected

Old: `SubscriberFetchModule::_update_subscribers function took 4342.029 ms`
New: `SubscriberFetchModule::_update_subscribers function took 1413.810 ms`

:+1:

File: pajbot/apiwrappers/twitch/helix.py [hunk 4]
```diff
@@ -347,7 +347,7 @@ def _fetch_subscribers_page(self, broadcaster_id: str, authorization, after_pagi
 
         return subscribers, pagination_cursor
 
-    def fetch_all_subscribers(self, broadcaster_id, authorization):
+    def fetch_all_subscribers(self, broadcaster_id: str, authorization) -> Set[UserBasics]:
```
Comment (RAnders00) [inline pajbot/apiwrappers/twitch/helix.py:352]: The function does not actually return a set!!

**Pull request merged**